### PR TITLE
[inductor] Cleanup strip_last_size logic

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100331
* #100308
* #100329
* __->__ #100305

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/dependencies.py [hunk 1]
```diff
@@ -220,10 +193,18 @@ def canonicalize(
...
+        new_vars, add_var = var_builder(canonicalization_prefix())
         replacement = dict(zip(index_vars, reindex([add_var(x) for x in new_sizes])))
-
         index = sympy_subs(sympy.expand(index), replacement)
+
+        new_vars = [*new_vars.keys()]
+        new_sizes = [*new_sizes]
+        free_symbols = index.free_symbols
+        while new_vars and new_vars[-1] not in free_symbols:
+            # Reduction has last (reduced) dim in its sizes, but
```
Comment (ngimel) [inline torch/_inductor/dependencies.py:204]: are there cases where downstream users still would have reduced dim? I remember I needed both versions to enable more fusions.

Comment (jansel) [reply torch/_inductor/dependencies.py:204]: I believe this logic will remove the extra dim from both the producer and the downstream users.

**Review by ngimel — APPROVED** [review]
Would be good to run perf test with this, to make sure it didn't break fusions.

**Comment by jansel** [comment]
I did spot test huggingface and didn't see any change in perf.

**Comment by jansel** [comment]
The reverted version of this actually triggered an extra `embedding+var_mean` fusion that tickled this Triton bug:
https://github.com/openai/triton/issues/1615

That one took a solid day to track down.  As a workaround I think we may need to disable inplace buffers.

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -1237,7 +1220,14 @@ def codegen(self):
...
             elif isinstance(node, (FusedSchedulerNode, SchedulerNode)):
-                self.get_backend(device).codegen_nodes(node.get_nodes())
+                with config.patch(
+                    inplace_buffers=(
+                        config.inplace_buffers
+                        # workaround https://github.com/openai/triton/issues/1615
+                        and not (ir.is_triton(device) and node.is_reduction())
+                    )
+                ):
+                    self.get_backend(device).codegen_nodes(node.get_nodes())
```
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/_inductor/scheduler.py:1230]: alternative is generating `tl.debug_barrier` as suggested on the issue, but it's more complicated so we can go with this.

**Pull request closed without merging**